### ISSUE TITLE
Channel architecture and mandated useEffect deps

### ISSUE THREAD

**Issue by nathanleclaire** [electron-react-boilerplate/electron-react-boilerplate#3113]
## Summary

I have a couple of questions about architecture and approach in ERB.

**Channels: How to structure them?**

So far I've been adding one channel to the `ipcRenderer` per 'endpoint', similar to how you'd have `/users`, `/accounts`, in a REST service. Is this the right approach or do you suggest consolidating them more and passing more properties in `msg`? e.g., I have separate `create-foo`, `delete-foo`, channels right now, seems best to roll them into one `foo` with a `method` parameter?

What's the right design for these?

** useEffect, once, deps, and lifecycle **

Probably due to my newbishness, I've had no end of confusion with how to handle the listeners vs. the calls to the main process. I have a variety of things where I want to load some data up once when a view loads, as well as a channel where the backend pushes many events to the frontend for a 'live' type view.

Originally I was mostly using `on` with a `useEffect` hook, but that got me into a lot of pickles. For starters, the usual get response, set state lifecycle was broken, so I ended up using workarounds with refs. ESlint complained about dep arrays, and if I tried to move my listener function outside of the `useEffect`, it complained about wanting `useCallback`, which I'm still not entirely sure I understand to be honest. So, when to use `useEffect`?

So, it's a mess. Lean more on `once` I'm thinking? Any other ideas?

** Cleanup: When to unsubscribe? **

Somewhat related to the previous question, but I'm having a hard time deciding when exactly I should be calling `removeListener`? The 'natural' place feels like the return function in `useEffect`, but that feels off for some reason.

**Comment by Karna98** [comment]
I started my project around 3 months back and referred to **ERB _ver. 4.1.0_** for project structure. 

Here's the link [Notes (stage-branch)](https://github.com/Karna98/Notes/tree/stage) to understand the approach.

Starting with **no. of channels**, I kept a **single-channel for IPC** (Refer [preload.js](https://github.com/Karna98/Notes/blob/stage/src/main/preload.js)).

Rather than creating separate channels based on endpoints (of application), it is easy to **pass the endpoint along with data (in a single object) to the Main process** and then process the data based on the endpoint and send the response to the renderer process.

Now, where to place the **listener on the renderer side**? Place it at the **topmost level of react app** and inside **useEffect with dependency array set to `[]`**. 
**_So how does this work?_**
When the application launches, the top-level Component will render only once and correspondingly listener will be initialized (only once since the dependency array is `[]`). The **listener will be active throughout the application**.

The above approach will help you start with IPC between the main and renderers process.

As we are setting up a single channel that needs to be active all the time, hence no need for a cleanup of listeners.

_**PS:** This is not a full-proof design and might evolve as problems start to arise._

**Comment by nathanleclaire** [comment]
@Karna98 thank you for the reply.

> Rather than creating separate channels based on endpoints (of application), it is easy to pass the endpoint along with data (in a single object) to the Main process and then process the data based on the endpoint and send the response to the renderer process.

Makes sense. Any particular reason for this? Simplicity, I suppose? I'm not loving adding a new channel for each endpoint, so I'm open to it, but curious if there is anything I'm missing.

> Now, where to place the listener on the renderer side? Place it at the topmost level of react app and inside useEffect with dependency array set to [].

This seems sensible, a couple questions on that -- do you run into ESLint complaining about things like using stuff from `useState` given that empty dep array? That's part of what seems to have started my whole ref clusterfuck. Also, I guess this means you are declaring just about all state and everything you use within the top level component and then passing it around everywhere?

I think I saw you're using Redux, perhaps that approach fits well with Redux's model? Can you talk a little about how that all ties in?

Thank you for your help!

**Comment by Karna98** [comment]
> Any particular reason for this? Simplicity, I suppose?  

Yes. The main aim was to keep it simple at the start and then expand it as the complexity increases. And one more reason is that I haven't come across any use-case for multi-channel IPC to get a better understanding of WHY and WHERE to use.

> do you run into ESLint complaining about things like using stuff from useState given that empty dep array? 

I assumed you meant `useEffect`. Yes, the ESlinst throws a warning when all the variables used in `useEffect` are not added to the dependency array. Since it is a warning, we can ignore it and the app will run without throwing any error.

> I guess this means you are declaring just about all states and everything you use within the top-level component and then passing it around everywhere?
I think I saw you're using Redux, perhaps that approach fits well with Redux's model? Can you talk a little about how that all ties in?

When the response is received from the Main process, I have created a custom response resolution hook ([Refer here](https://github.com/Karna98/Notes/blob/stage/src/renderer/Hooks/useResponse.ts)) to identify the response and accordingly store it in a centralized store. So, the centralized store is `redux store` and a better version of it is  `redux-tookit`. The main reason for using `redux` is to avoid passing data through components from parent to children and grandchildren. And if we are going to use react then we don't have an idea how deep the hierarchy can be so using `redux store` look like a good option at this moment.

**_PS;_** The **Notes** application is an offline application.

**Comment by nathanleclaire** [comment]
That definitely confirms what I was thinking, thanks!

**Comment by amilajack** [comment]
This seems resolved. Closing but feel free to continue the convo

**Comment by sanneh2** [comment]
How do you register removeAllListeners to ipcrenderer in preload.js? Just adding it doesn't work as it doesn't register this function. 
I am using multiple channels for ipc and the listeners are registered on component load.
I wanted to unregister the listeners of course, but got window.electron.ipcRenderer.removeAllListeners is not a function.
